pkg/policy/selectorcache.go

```diff
@@ -0,0 +1,336 @@
+// depending on what selectors the rules contain. Cached selections of
+// new IdentitySelectors are pre-populated from the set of currently
+// known identities.
+//
+// 2. When reachacble identities appear or disappear, either via local
+// allocation (CIDRs), or via the KV-store (remote endpoints). In this
+// case all existing IdentitySelectors are walked through and their
+// cached selections are updated as necessary.
+//
+// In both of the above cases the set of existing IdentitySelectors is
+// write locked.
+//
+// To minimize the upkeep the identity selectors are shared accross
+// all IdentityPolicies, so that only one copy exists for each
+// IdentitySelector. Users of the SelectorCache take care of creating
+// IdentitySelectors as needed by identity policies. The set of
+// IdentitySelectors is read locked during an IdentityPolicy update so
+// that the the policy is always updated using a coherent set of
+// cached selections.
+//
+// IdentitySelector is used as a map key, so it must not be implemented by a
+// map, slice, or a func, or a runtime panic will be triggered. In all
+// cases below IdentitySelector is being implemented by structs.
+type IdentitySelector interface {
+	CachedSelector
+	addUser(CachedSelectionUser) (added bool)
+	removeUser(CachedSelectionUser) (last bool)
+	notifyUsers(added, deleted []identity.NumericIdentity)
+}
+
+// SelectorCache caches identities, identity selectors, and the
+// subsets of identities each selector selects.
+type SelectorCache struct {
+	mutex lock.RWMutex
+
+	// idCache contains all known identities as informed by the
+	// kv-store and the local identity facility via our
+	// UpdateIdentities() function.
+	idCache cache.IdentityCache
+
+	// map key is the string representation of the selector being cached.
+	selectors map[string]IdentitySelector
+}
+
+// NewSelectorCache creates a new SelectorCache.
+func NewSelectorCache() *SelectorCache {
+	return &SelectorCache{
+		idCache:   cache.GetIdentityCache(),
+		selectors: make(map[string]IdentitySelector),
+	}
+}
+
+var (
+	// selectorCache is the global selector cache. Additional ones
+	// are only created for testing.
+	selectorCache  = NewSelectorCache()
+	emptySelection = make([]identity.NumericIdentity, 0)
+)
+
+type labelIdentitySelector struct {
+	selector         api.EndpointSelector
+	key              string
+	users            map[CachedSelectionUser]struct{}
+	selections       unsafe.Pointer
+	cachedSelections map[identity.NumericIdentity]struct{}
+}
+
+//
+// CachedSelector implementation (== Public API)
+//
+// No locking needed.
+//
+
+// GetSelections returns the set of numeric identities currently
+// selected.  The cached selections can be concurrently updated. In
+// that case GetSelections() will return either the old or new version
+// of the selections. If the old version is returned, the user is
+// guaranteed to receive a notification including the update.
```

Comment:
Right, there will be a notification of for the change that caused the new version to be created, but as updates are coalesced by the next time the user calls `GetSelections()` it may already be updated with further changes.